File: web/src/audio/dsp.js
/**
 * Framing and spectrum — the JS side of spike/core/dsp.py.
 *
 * Must reproduce numpy's `abs(rfft(x * hann))` to within 1e-6. Two things carry
 * almost all the port risk:
 *
 *  1. The Hann window is PERIODIC (divisor N), not symmetric (divisor N-1).
 *     numpy.hanning gives the symmetric form. Getting this wrong shifts every
 *     magnitude slightly and silently invalidates every tuned threshold.
 *
 *  2. numpy's rfft is UNSCALED — no 1/N anywhere. Many JS FFT implementations
 *     normalise the forward transform. This one does not.
 *
 * Both are pinned by test/golden.test.js.
 */

const windowCache = new Map();

/** Periodic Hann: 0.5 - 0.5*cos(2*pi*k/n), k in [0, n). Divisor n, not n-1. */
export function hannPeriodic(n) {
  let w = windowCache.get(n);
  if (!w) {
    w = new Float64Array(n);
    for (let k = 0; k < n; k++) w[k] = 0.5 - 0.5 * Math.cos((2 * Math.PI * k) / n);
    windowCache.set(n, w);
  }
  return w;
}

const twiddleCache = new Map();

function twiddles(n) {
  let t = twiddleCache.get(n);
  if (!t) {
    const cos = new Float64Array(n / 2);
    const sin = new Float64Array(n / 2);
    for (let i = 0; i < n / 2; i++) {
      cos[i] = Math.cos((-2 * Math.PI * i) / n);
      sin[i] = Math.sin((-2 * Math.PI * i) / n);
    }
    t = { cos, sin };
    twiddleCache.set(n, t);
  }
  return t;
}

/**
 * In-place iterative radix-2 Cooley-Tukey FFT, unscaled forward transform.
 * `re` and `im` are Float64Array of length n, n a power of two.
 */
export function fftInPlace(re, im) {
  const n = re.length;
  if (n !== im.length) throw new Error('re/im length mismatch');
  if ((n & (n - 1)) !== 0) throw new Error(`fft size ${n} is not a power of two`);

  // Bit-reversal permutation.
  for (let i = 1, j = 0; i < n; i++) {
    let bit = n >> 1;
    for (; j & bit; bit >>= 1) j ^= bit;
    j ^= bit;
    if (i < j) {
      let t = re[i]; re[i] = re[j]; re[j] = t;
      t = im[i]; im[i] = im[j]; im[j] = t;
    }
  }

  const { cos, sin } = twiddles(n);
  for (let len = 2; len <= n; len <<= 1) {
    const step = n / len;
    const half = len >> 1;
    for (let i = 0; i < n; i += len) {
      for (let k = 0; k < half; k++) {
        const tw = k * step;
        const wr = cos[tw];
        const wi = sin[tw];
        const a = i + k;
        const b = a + half;
        const xr = re[b] * wr - im[b] * wi;
        const xi = re[b] * wi + im[b] * wr;
        re[b] = re[a] - xr;
        im[b] = im[a] - xi;
        re[a] += xr;
        im[a] += xi;
      }
    }
  }
}

/**
 * Windowed magnitude spectrum of one frame: fftSize/2 + 1 non-negative values.
 *
 * Magnitudes, not power — chroma sums them linearly, and squaring would
 * over-weight the loudest partial and change what the harmonic weights mean.
 */
export function spectrum(frame, fftSize, scratch) {
  const n = fftSize ?? frame.length;
  if (frame.length !== n) throw new Error(`frame length ${frame.length} != fftSize ${n}`);

  const re = scratch?.re ?? new Float64Array(n);
  const im = scratch?.im ?? new Float64Array(n);
  const out = scratch?.mag ?? new Float64Array(n / 2 + 1);

  const w = hannPeriodic(n);
  for (let i = 0; i < n; i++) {
    re[i] = frame[i] * w[i];
    im[i] = 0;
  }

  fftInPlace(re, im);

  for (let k = 0; k <= n / 2; k++) out[k] = Math.hypot(re[k], im[k]);
  return out;
}

/** Reusable buffers, so the per-frame path allocates nothing. */
export function makeScratch(fftSize) {
  return {
    re: new Float64Array(fftSize),
    im: new Float64Array(fftSize),
    mag: new Float64Array(fftSize / 2 + 1),
  };
}

export function rms(frame) {
  if (frame.length === 0) return 0;
  let sum = 0;
  for (let i = 0; i < frame.length; i++) sum += frame[i] * frame[i];
  return Math.sqrt(sum / frame.length);
}

export function clippedFraction(samples, threshold) {
  if (samples.length === 0) return 0;
  let n = 0;
  for (let i = 0; i < samples.length; i++) if (Math.abs(samples[i]) >= threshold) n++;
  return n / samples.length;
}

File: spike/core/dsp.py
"""Framing and spectrum. The entire DSP surface of the spike is in this file.

Kept to `numpy.fft.rfft` and an explicitly-written Hann window so that the
browser port is a transcription rather than a reimplementation (DECISION #16).

One subtlety worth naming, because it is the classic source of port drift:
`numpy.hanning(N)` is the *symmetric* window (divisor N-1), while the periodic
window (divisor N) is what STFT analysis wants and what every JS FFT example
uses. We build the periodic form by hand below. A symmetric/periodic mismatch
between this file and the JS port would shift every magnitude slightly and
quietly invalidate the tuned thresholds.
"""

from __future__ import annotations

import numpy as np

_WINDOW_CACHE: dict[int, np.ndarray] = {}


def hann_periodic(n: int) -> np.ndarray:
    """Periodic Hann window: ``0.5 - 0.5 * cos(2*pi*k/n)`` for k in [0, n).

    Note the divisor is `n`, not `n - 1`. See the module docstring.
    """
    cached = _WINDOW_CACHE.get(n)
    if cached is None:
        k = np.arange(n, dtype=np.float64)
        cached = 0.5 - 0.5 * np.cos(2.0 * np.pi * k / n)
        _WINDOW_CACHE[n] = cached
    return cached


def frame_count(num_samples: int, fft_size: int, hop: int) -> int:
    """How many whole frames fit in a signal of this length."""
    if num_samples < fft_size:
        return 0
    return 1 + (num_samples - fft_size) // hop


def iter_frames(samples: np.ndarray, fft_size: int, hop: int):
    """Yield successive un-windowed frames of `fft_size` samples.

    Trailing samples that do not fill a whole frame are dropped rather than
    zero-padded: a partial frame has a different effective window and would
    produce a magnitude spectrum that is not comparable with its neighbours.
    """
    for i in range(frame_count(len(samples), fft_size, hop)):
        start = i * hop
        yield samples[start : start + fft_size]


def spectrum(frame: np.ndarray, fft_size: int | None = None) -> np.ndarray:
    """Windowed magnitude spectrum of one frame.

    Returns ``fft_size // 2 + 1`` non-negative magnitudes (not power, not dB).
    Magnitudes are used throughout because chroma sums them linearly; squaring
    would over-weight the loudest partial and make the fundamental-vs-harmonic
    weighting in `chroma.py` mean something different than intended.
    """
    n = fft_size if fft_size is not None else len(frame)
    if len(frame) != n:
        raise ValueError(f"frame length {len(frame)} != fft_size {n}")
    return np.abs(np.fft.rfft(frame * hann_periodic(n)))


def bin_frequencies(fft_size: int, sample_rate: int) -> np.ndarray:
    """Centre frequency of each rfft bin, in Hz."""
    return np.fft.rfftfreq(fft_size, d=1.0 / sample_rate)


def rms(frame: np.ndarray) -> float:
    """Root-mean-square level of a frame, on the same scale as the samples
    (so roughly 0.0 to 1.0 for normalised float audio)."""
    if len(frame) == 0:
        return 0.0
    return float(np.sqrt(np.mean(np.square(frame, dtype=np.float64))))


def clipped_fraction(samples: np.ndarray, threshold: float) -> float:
    """Fraction of samples at or beyond `threshold` in absolute value.

    Surfaced everywhere because a clipped recording generates harmonic
    distortion that looks exactly like extra notes — it would corrupt the
    extra-note rejection measurement while appearing to be a real result.
    """
    if len(samples) == 0:
        return 0.0
    return float(np.mean(np.abs(samples) >= threshold))
